content/blog/shop-and-thumbnails.md

```diff
@@ -0,0 +1,77 @@
+---
+id: shopware-and-thumbnails
+date: 2021-07-05T23:15:59
+title: Thumbnails + Shopware + PHP = 🙉
+author: Shyim
+author_title: Developer @ Shopware
+author_url: https://github.com/shyim
+author_image_url: https://avatars3.githubusercontent.com/u/6224096?s=460&u=18be3a2d46f07dd42fc2b6dee9b4b9b68bca28d2&v=4
+tags: [shopware]
+---
+
+All shop owners want a fast online shop with the best pagespeed results. The images in the shop are making a big difference in this stats.
+But can we archive this with only PHP?
+
+## Image generation with PHP
+
+In PHP there are two ways how images can be generated: using the `gd`, or the `imagemagic` extension. Both of these extensions are not optimized to create very small images and they are pretty slow. They also lack support for newer image formats like avif (PHP 8.1 will have avif in `gd`).
+The PHP community has build packages like [spatie/image-optimizer](https://github.com/spatie/image-optimizer) to optimize an image after it has been uploaded by executing several optimziers in the CLI. This approach requires to have functions like `exec`, `system` or `proc_open` available, which are disabled by default on some shared hosters. Also, the optimizers need to be installed on the host system. But there are new technologies like `ffi` or `wasm` available in PHP.
+
+## Image generation using FFI / WASM
+
+Instead of requiring optimizers installed on the server system, we could bring it with the code, isolated as WASM or as a shared library with FFI.
+These approaches are both very fast, but unfornuately both technologies are not widespread in the PHP ecosystem.
+
+## Letting a external service do your image processing
+
+When we can't create the best optimized image with just plain PHP, why not let our image processing work be done by some external service? One of my favourite image processing systems is [imgproxy](https://imgproxy.net/). It can be even selfhosted, if necessary. We save the original images on the server or an external storage like S3 and just link it to our image processing server with some parameters like width, height etc. The servers are doing the rest for us with the best possible image results.
+Such servers are also offering to provide better image formats when the browser supports it. Like when the browser is requesting a `jpg` image, but the browser supports also `webp`. The server will respond to the `.jpg` file extension a `webp` image without even providing multiple links or using a `picture` element.
+**It's important** that your used processing service does cache your requests or you will have to setup a own CDN before it.
+Let's do a quick `pro` and `contra` list of the usage of such software:
```

Comment:
```suggestion
When we can't create the best optimized image with just plain PHP, why not let our image processing work be done by some external service? One of my favourite image processing systems is [imgproxy](https://imgproxy.net/). It can be even selfhosted, if necessary. We save the original images on the server or an external storage like S3 and just link it to our image processing server with some parameters, like width, height etc. The servers are doing the rest for us with the best possible image results.
Such servers also offer better image formats if the browser supports it. This means, when the browser requests a `jpg` image, but the browser supports also `webp`. The server can respond to the `.jpg` file extension a `webp` image without even providing multiple links or using a [`picture` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture#the_type_attribute).
**It's important** that your used processing service does cache your requests or you will have to setup a own CDN before it.
Let's do a quick `pros` and `cons` list of the usage of such software:
```